### PR TITLE
feat(plugin): add build plugin via unplugin

### PR DESCRIPTION
**Pull request by JonasBa**
Opening as draft and looking for early feedback - I want to add actual tests for every plugin and just generally looking for opinions on the PR. Imo the logic is simple enough where this doesn't warrant a separate package, but happy to split this out @markerikson

### CODE REVIEW

File: package-lock.json
Comment (markerikson) [inline package-lock.json:1]: :hand: Stick with Yarn 1 for deps for now, please :)

**Review by markerikson — COMMENTED** [review]
Changes seem reasonable at first glance!

**Review by jasikpark — COMMENTED** [review]
looks reasonable to me, from a quick skim

File: package-lock.json
Comment (JonasBa) [reply package-lock.json:1]: Argh, yes sorry about that :) will update the PR and add some tests before I ask for another review

**Comment by markerikson** [comment]
Did some work to try this out with our actual Replay frontend at https://github.com/replayio/devtools , which is a Next app.

Found a couple typos in the plugin exports here, and I ran into issues where the filename reported by `remapping` was:

```
webpack://_N_E/node_modules/react-dom/cjs/react-dom.production.min.js
```

but the filename in the sources content was:

```
webpack://_N_E/./node_modules/react-dom/cjs/react-dom.production.min.js
```

Looks like `trace-mapping` does a resolve on the path segments.

Updated the logic to do the same resolve comparison as a fallback, and it works!  Here's a screenshot of the `framework-*.js` sourcemap from the Next build:

image

Gonna merge this and put out 0.0.3 so I can actually try it with a PR to Replay!  Please feel free to file follow-up PRs with tests and such.

**Comment by markerikson** [comment]
Okay, quick update:

- published `0.0.4` with a fix to more strictly search for `"react-dom.production.min"`, to avoid the second pass where `remapping` lists our `react-dom.production.js` replacement filename (and also tried to handle the plugin options not existing)
- Put up https://github.com/replayio/devtools/pull/9706 to add it to Replay...

AND IT WORKS!

Uh, mostly.

The actual sourcemapping replacement step works.  But the sourcemap itself is a bit wonky.  Function arguments don't appear to be getting mapped right:

image

that definitely reduces usability.

Going to have to investigate further and see if that's coming straight from Closure Compiler, or if something else is going on.

BUT HEY, look, I can at least see valid hits in a Replay recording of that PR branch!

image

**Comment by markerikson** [comment]
Oh drat.  This is a _Next_ problem, isn't it!

I'm looking at one of my generated sourcemaps, and _it_ pretty clearly has per-argument mappings:

image

but the sourcemap for the entire Next app doesn't - functions seem to get smooshed together.

My teammate Holger has an open bug against Next complaining about something similar, and I wonder if that's a problem here:

- https://github.com/vercel/next.js/issues/53253

**Comment by JonasBa** [comment]
@markerikson thank you so much for the investigation! I was going to remove the 2nd pass on the remapping, but wasn't entirely sure if there were edge cases I'm not considering and hence why you opted for that behavior. I will ask our team internally if they are aware of the inaccurate sourcemap problems in next.

File: src/index.ts [hunk 1]
```diff
@@ -46,20 +58,34 @@ export function loadSourcemap(filePath: string) {
...
 function findMatchingReactDOMVersion(
   reactDomFilename: string,
   inputSourcemap: SourceMapV3
 ): ReactVersion {
-  const filenameIndex = inputSourcemap.sources.indexOf(reactDomFilename);
+  let filenameIndex = inputSourcemap.sources.indexOf(reactDomFilename);
+  if (filenameIndex === -1) {
+    // Try one more time. Maybe the path had extra segments in it, like:
+    // "webpack://_N_E/./node_modules/react-dom/cjs/react-dom.production.min.js"
+    filenameIndex = inputSourcemap.sources.findIndex(filename => {
```
Comment (JonasBa) [inline src/index.ts:82]: Good call 👍🏼

File: src/unplugin.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
+import * as fs from "node:fs";
+import * as path from "node:path";
+import { log } from "console";
+import { createUnplugin } from "unplugin";
+import { maybeRewriteSourcemapWithReactProd, loadSourcemap } from "./index";
+
+export interface ReactSourcemapsPluginOptions {
```
Comment (JonasBa) [inline src/unplugin.ts:7]: I wanted to also add a "preserve" option which would preserve the original sourcemaps with a original.js.map suffix, wdyt @markerikson ?

Comment (markerikson) [reply src/unplugin.ts:7]: Seems reasonable, sure!

**Pull request merged**